### DISCUSSION TITLE
生存分析在case-cohort中使用权重的方法

### DISCUSSION THREAD

**Discussion by JYL2303** [ShixiangWang/self-study#43] [Q&A] [answered]
您好，有个问题想请教一下。我是在一篇文章中发现的一个统计分析方法，只知道具体是使用生存分析的cox模型，加了权重。
我只查到大概是这样，但是我不确定是否是我自己理解的那样。因为数据结果对不上。
具体文章的统计方法是这样描述的：
After proper exclusion, each patient was assigned weights, 1 for cases and N/n for non-cases in the sub-cohort, with N being the number of non-cases in the cohort and n being the number of non-cases in the sub-cohort. 
这是我自己查到的R代码：
setotal$weights.trunc <- with(setotal,ifelse(stroke==1,1,ifelse(stroke==0,N/n,NA)))
HR5 <- coxph(Surv(time, as.numeric(stroke))~se.3t, weights = weights.trunc, data = setotal)
这是我查到的理论方法：
![image](https://user-images.githubusercontent.com/54757508/86774295-febb6280-c088-11ea-82c2-92e734c14295.png)

万分感谢，祝好！

**Comment by ShixiangWang** [comment]
中文的表述我不是很理解，我没有特别深的统计背景。从英文描述来看，应该是这样给病人赋予权重：

- 如果这个病人死亡（即所谓的case事件），那么权重为1
- 如果这个病人没有死亡（即non-case事件），那么权重为 N/n，N 为整个数据集的病人数，而n为这个病人所属子队列的病人数

另外模型中为啥 status 是 as.numeric(stroke)，是因为 stroke 是表示事件发生了吗？如果不是的话，可能生存对象是有问题的。

**Comment by JYL2303** [comment]
“如果这个病人没有死亡（即non-case事件），那么权重为 N/n，N 为整个数据集的病人数，而n为这个病人所属子队列的病人数”
这里是否应该是N 为整个non-case数据集的人数，而n为这个病人所属non-case子队列的人数？
stroke是0和1的变量，1是发生了终点事件，0是未发生终点事件。我一开始设置这个变量是因子，但是报错了，变成数字型又可以了，请问这里我是不是理解上有错误？
总体上coxph的代码我是没有用错的是吗？这个代码单独跑没有问题，但是放进function里面总是出现报错，不放weights的话都没有问题，但是一旦放了weight，function就报错了，具体是这样的：

这是单独跑的：
> setotal$weights.trunc <- with(setotal,ifelse(stroke==1,1,ifelse(stroke==0,nrow(sestrokebsw)/nrow(sestroke15002),NA)))
> HR5 <- coxph(Surv(stroke.time, as.numeric(stroke))~se.sd, weights = weights.trunc, data = setotal)
> (X <- tidy(HR5)%>% .[1,c(2:3,5:7)])
 A tibble: 1 x 5
  estimate std.error p.value conf.low conf.high
     <dbl>     <dbl>   <dbl>    <dbl>     <dbl>
1   -0.113    0.0429 0.00837   -0.197   -0.0290

这是放进函数里面的：
> coxcon.f <- function(db,y,x,time,weights) {
   attach(db)
   cru<-tidy(coxph(Surv(time, y)~x, weights)) %>% 
     .[1,c(2:3,5:7)]
   detach(db)
   return(cru)
 }
> coxcon.f(db = setotal,time = stroke.time,y = as.numeric(stroke),x = se.sd, weights=weights.trunc)
 Error in eval(predvars, data, env) : 数值'envir'参数的长度不是一

**Comment by ShixiangWang** [comment]
前面有一点没说对
N 为整个数据集的没有死亡病人数，而n为这个病人所属子队列的没有死亡病人数

你单独运行是否与预期结果一致？

函数报错是因为你代码有问题，你先需要确定分析是否是对的

**Comment by JYL2303** [comment]
单独运行与预期有点偏差，HR与预期一样，但是P值和置信区间有点偏差。对于weights的理解上的偏差，对于N和n的理解。
以下是完整版本的统计方法分析：
Statistical Analyses
Data are presented as mean ± standard deviation (SD) for continuous variables and as frequency (%) for categorical variables. The population characteristics of the different groups were compared using ANOVA tests, or chi-square tests, accordingly. After proper exclusion, each patient was assigned weights, 1 for cases and N/n for non-cases in the sub-cohort, with N being the number of non-cases in the cohort and n being the number of non-cases in the sub-cohort. The hazard ratios (HRs) and 95% confidence intervals (CI) for the risk of a primary outcome associated with plasma selenium levels as a continuous variable (each standard deviation increment) and as tertiles were estimated using logistic regression without and with adjustment for age, sex, treatment groups, study centers, body-mass index (BMI), smoking status, alcohol drinking, MTHFR C677T polymorphism, total cholesterol (TC), triglycerides (TG), high-density lipoprotein cholesterol (HDL-C), fasting glucose, estimated glomerular filtration rate (eGFR), total homocysteine (tHcy), folate, vitamin B12, systolic blood pressure (SBP) and diastolic blood pressure (DBP) at baseline, as well as SBP and DBP during the treatment period. In addition, possible modifications of the association between selenium and stroke were also assessed. A two-tailed P<0.05 was considered to be statistically significant in all analyses.

**Comment by ShixiangWang** [comment]
这个文章里使用的是多变量分析，你做的是单变量？？具体原因我说不清楚

函数的构建你需要理由字符串构建公式，请参考 https://openbiox.github.io/Cookbook-for-R-Chinese/section-4.html#section-4.1

**Comment by JYL2303** [comment]
是多变量分析，我只是先用单变量看看是否能对的上。
十分感谢，我去试一下。

**Comment by JYL2303** [comment]
下午好。
我试着用字符串构建公式，也出了问题，烦请赐教：
measurevar <- "y"
groupvars <- "x"
w <- "weights"
S <- "Surv"
Tm <- "time"
C <- "coxph"
(p1.1 <- paste(C, "(", S, "(",Tm , ",", measurevar , ")" ,sep = ""))
(p1.2 <- paste(groupvars, ",", w , ")",sep = ""))
(p1.3 <- paste(p1.1, p1.2, sep = "~"))
(f1 <- as.formula(p1.3))
Error in terms.formula(formula, specials = ss, data = data) : 
  'data'参数种类有错


加上data也报错：
data <- "data"
(p1.1 <- paste(C, "(", S, "(",Tm , ",", measurevar , ")" ,sep = ""))
(p1.2 <- paste(groupvars, ",", w, ",", data , ")",sep = ""))
(p1.3 <- paste(p1.1, p1.2, sep = "~"))
(f1 <- as.formula(p1.3))
Error in terms.formula(formula, specials = ss, data = data) : 
  'data'参数种类有错

**Comment by ShixiangWang** [comment]
不好意思很久才回，这个 coxph 是函数，并不是公式的一部分啊~

**Comment by ShixiangWang** [comment]
这是一个示例：

```
> library(survival)
> coxph(as.formula("Surv(time, status) ~ sex"), data = lung)
Call:
coxph(formula = as.formula("Surv(time, status) ~ sex"), data = lung)

       coef exp(coef) se(coef)      z       p
sex -0.5310    0.5880   0.1672 -3.176 0.00149

Likelihood ratio test=10.63  on 1 df, p=0.001111
n= 228, number of events= 165 
```

字符串部分你是可自动拼接的。如果你不太会生存分析，可以使用我写的包 https://cran.r-project.org/web/packages/ezcox/index.html

**Comment by JYL2303** [comment]
没事，理解的。您可以抽时间回答我的问题已经万分感谢了！
关于您的R包，我已经下载，感谢分享，很实用。但是我想要加上weights权重部分。好像没有看到。
coxcon.f <- function(db,y,x,time,weights) {
  attach(db)
  cru<-tidy(coxph(as.formula("Surv(time, as.numeric(y)) ~ x,weights"))) %>% .[1,c(2:3,5:7)]
  detach(db)
  return(cru)
}
coxcon.f(db = setotal,time = stroke.time,y = stroke,x = se.sd, weights=weights.trunc)

 Error in parse(text = x, keep.source = FALSE) : <text>:1:30: 意外的','
1: Surv(time, as.numeric(y)) ~ x,
                                                   ^ 

不加weghts是没有问题的，但是一加就老是说多了“，”，不是太明白。感谢解惑~

**Comment by ShixiangWang** [comment, marked as answer]
weights 不是公式的一部分，它是 coxph 函数的一个参数，我随便举个例子

```r
> coxph(as.formula("Surv(time, status) ~ sex"), data = lung, weights = rep(0.5, length(lung$sex)))
Call:
coxph(formula = as.formula("Surv(time, status) ~ sex"), data = lung, 
    weights = rep(0.5, length(lung$sex)))

       coef exp(coef) se(coef) robust se      z        p
sex -0.5310    0.5880   0.2364    0.1600 -3.318 0.000906

Likelihood ratio test=5.32  on 1 df, p=0.02112
n= 228, number of events= 165 
```

**Comment by JYL2303** [comment]
明白了！非常非常感谢！总是能在您身上学到很多东西，万分感谢！
